File: 797_all_paths_from_source_to_target.java
class Solution {
    // apply dfs visiting
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        path.add(0);
        findPath(0, res, path, graph);
        return res;
    }
    
    private void findPath(int curr, List<List<Integer>> res, List<Integer> temp, int[][] graph) {
        if (curr == graph.length - 1) {
            res.add(new ArrayList<>(temp));
            return;
        }
        for (int neighbor: graph[curr]) {
            temp.add(neighbor);
            findPath(neighbor, res, temp, graph);
            temp.remove(temp.size() - 1);
        }
    }
}

File: python/797_all_paths_from_source_to_target.py
class Solution:
    # Time Complexity: O(2 ^ N)
    # Space Complexity: O(2 ^ N)
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def dfs(cur, path):
            if cur == len(graph) - 1: res.append(path)
            else:
                for i in graph[cur]:
                    dfs(i, path + [i])
        res = []
        dfs(0, [0])
        return res